apps/constellation/key_generator.cpp

```diff
@@ -0,0 +1,108 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "crypto/ecdsa.hpp"
+
+#include <cstdlib>
+#include <fstream>
+
+namespace fetch {
+
+using fetch::crypto::Prover;
+using fetch::crypto::ECDSASigner;
+
+using SignerPtr = std::shared_ptr<ECDSASigner>;
+using ProverPtr = std::shared_ptr<Prover>;
+
+char const *LOGGING_NAME         = "KeyGenerator";
+char const *DEFAULT_KEY_FILENAME = "p2p.key";
+
+/**
+ * Lookup the key path for the key
+ *
+ * @return The path to the key file
+ */
+char const *GetKeyPath()
+{
+  char const *path = std::getenv("CONSTELLATION_KEY_PATH");
+
+  if (path == nullptr)
+  {
+    path = DEFAULT_KEY_FILENAME;
+  }
+
+  return path;
+}
+
+/**
+ * Attempt to load a previous key key file
+ *
+ * @return The shared pointer to the new key
+ */
+ProverPtr GenerateP2PKey()
+{
+  char const *key_path = GetKeyPath();
+
+  SignerPtr certificate        = std::make_shared<ECDSASigner>();
+  bool      certificate_loaded = false;
+
+  // Step 1. Attempt to load the existing key
+  {
+    std::ifstream input_file(key_path, std::ios::in | std::ios::binary);
+
+    if (input_file.is_open())
+    {
+      fetch::byte_array::ByteArray private_key_data;
+      private_key_data.Resize(ECDSASigner::PrivateKey::ecdsa_curve_type::privateKeySize);
```

Comment:
I think ByteArray has a ctor which takes a size_t argument and calls resize for you

Reply:
Nope that is not the case (unfortunately)

Reply:
Seems like it would be useful. Consider adding it as part of this change.

Reply:
Seems like it would be useful. Consider adding it as part of this change?

Reply:
Not for this change. It would be a change in of itself since this pattern of create instance and resize is prevalent in the code base

Reply:
Fair. Please create a Jira issue - if the pattern is this prevalent, it'll be a nice bit of refactoring to do at some point.